src/Native/Runtime/unix/UnixEvent.cpp

```diff
@@ -0,0 +1,334 @@
+
+#endif // HAVE_CLOCK_MONOTONIC
+    }
+
+    int st = 0;
+
+    pthread_mutex_lock(&m_mutex);
+    while (m_eventIndex == NoEventIndex)
+    {
+        if (milliseconds == INFINITE)
+        {
+            st = pthread_cond_wait(&m_condition, &m_mutex);
+        }
+        else
+        {
+#ifdef __APPLE__
+            // Since OSX doesn't support CLOCK_MONOTONIC, we use relative variant of the 
+            // timed wait and we need to handle spurious wakeups properly.
+            st = pthread_cond_timedwait_relative_np(&m_condition, &m_mutex, &endTime);
+            if ((st == 0) && (m_eventIndex != NoEventIndex))
+            {
+                uint64_t machTime = mach_absolute_time();
+                if (machTime < endMachTime)
+                {
+                    // The wake up was spurious, recalculate the relative endTime
+                    uint64_t remainingNanoseconds = (endMachTime - machTime) * timeBaseInfo.numer / timeBaseInfo.denom;
+                    NanosecondsToTimespec(remainingNanoseconds, &endTime);
+                }
+                else
+                {
+                    // Although the timed wait didn't report a timeout, time calculated from the
+                    // mach time shows we have already reached the end time. It can happen if
+                    // the wait was spuriously woken up right before the timeout.
+                    st = ETIMEDOUT;
+                }
+            }
+#else // __APPLE__ 
+            st = pthread_cond_timedwait(&m_condition, &m_mutex, &endTime);
+            // Verify that if the wait timed out, the event was not set
+            ASSERT((st != ETIMEDOUT) || (m_eventIndex == NoEventIndex));
+#endif // __APPLE__
+        }
+
+        if (st != 0)
+        {
+            m_timedOutOrFailed = true;
+            break;
+        }
+    }
+
+    pthread_mutex_unlock(&m_mutex);
+
+    uint32_t waitStatus;
+
+    if (st == 0)
+    {
+    	ASSERT(m_eventIndex != NoEventIndex);
+        waitStatus = WAIT_OBJECT_0 + m_eventIndex;
+    }
+    else if (st == ETIMEDOUT)
+    {
+        waitStatus = WAIT_TIMEOUT;
+    }
+    else
+    {
+        waitStatus = WAIT_FAILED;
+    }
+    return waitStatus;
+}
+
+// Signal a waiter that the event was set
+// Parameters:
+//  index  - index of the current event in the list of events associated with the waiter
+//  waiter - the waiter to signal
+void UnixEvent::SignalWaiter(int index, UnixEventWaiter* waiter)
+{
+    if (waiter->Signal(index) && !m_manualReset)
+    {
+    	// Autoreset event released a waiter, so we need to clear
+    	// the event state
+    	m_state = false;
+    }	
+}
+
+// Set the event to signalled state
+void UnixEvent::Set()
+{
+    pthread_mutex_lock(&m_mutex);
+    // Signal waiters only when the event transitions from not set to set.
+    if (!m_state)
+    {
+        m_state = true;
+
+        // Pass the event to all waiters for manual reset event
+        // or to the first waiter for autoreset event.
```

Comment:
It does, the m_state is reset after the first waiter for the autoreset event is signalled and we exit the loop.